### DISCUSSION TITLE
Asynchronous transactional operations cannot be saved.

### DISCUSSION THREAD

**Discussion by beimusky** [sqlalchemy/sqlalchemy#10621] [Usage Questions] [answered]
```py
SQLALCHEMY_URI = 'mysql+aiomysql://jdk_service:D234aaz_2@127.0.0.1:3306/apidb?charset=utf8mb4'


async def fun1(async_session):
    async with async_session() as session:
        result = await session.execute(select(User).order_by(JdkUser.id).limit(1))
        a1 = result.scalars().one()
        a1.password = "aaaaaa"


async def fun2(
        async_session: async_sessionmaker[AsyncSession],
) -> None:
    async with async_session() as session:
        result = await session.execute(select(User).order_by(JdkUser.id).limit(1))
        a2 = result.scalars().one()
        a2.nickname = "new data"


async def async_main() -> None:
    engine = create_async_engine(
        SQLALCHEMY_URI,
        echo=True,
    )
    async_session = async_sessionmaker(engine, expire_on_commit=False)
    await fun1(async_session)
    await fun2(async_session)
    await async_session().commit()
    await engine.dispose()


asyncio.run(async_main())
```

After running the code, there were no errors. However, the data did not change in the database. But adding commit inside the function can save the changes. If I want to perform asynchronous transaction operations, how should I do it?

**Comment by CaselIT** [comment, marked as answer]
Hi

That's the expected behavior of `async_sessionmaker`: by default a context manager will just close the session, without doing any commit.
Use `begin()` to create a context that will automatically commit at the end if no exception are raised: https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#sqlalchemy.ext.asyncio.async_sessionmaker.begin


Also note that each function will have a different session, so commands like `async_session().commit()` have no effect, since you are creating a new session and the committing, but not command was emitted in that session
